src/trigger/TriggerGenericMaker.hpp

```diff
@@ -72,42 +115,225 @@ class TriggerGenericMaker : public dunedaq::appfwk::DAQModule
+      ers::warning(excpt);
+      return false;
+    }
+    ++m_sent_count;
+    return true;
+  }
+  
+};
+
+// To handle the different unpacking schemes implied by different templates,
+// do_work is broken out into its own template class that is a friend of 
+// TriggerGenericMaker. C++ still does not support partial specification of a 
+// single method in a template class, so this approach is the least redundant
+// way to achieve that functionality
 
+// The base template assumes the MAKER has an operator() with the signature
+// operator()(IN, std::vector<OUT>)
+template<class IN, class OUT, class MAKER>
+class TriggerGenericWorker
+{
+public:
+  explicit TriggerGenericWorker(TriggerGenericMaker<IN, OUT, MAKER> &parent) : m_parent(parent) { }
+  
+  TriggerGenericMaker<IN, OUT, MAKER> &m_parent;
+  
+  void do_work(std::atomic<bool> &running_flag)
+  {
     while (running_flag.load()) {
       IN in;
+      if (!m_parent.receive(in)) {
+        continue; //nothing to do
+      }
+  
+      std::vector<OUT> out_vec; // 1 input -> many outputs
       try {
-        m_input_queue->pop(in, m_queue_timeout);
-      } catch (const dunedaq::appfwk::QueueTimeoutExpired& excpt) {
-        // it is perfectly reasonable that there might be no data in the queue
-        // some fraction of the times that we check, so we just continue on and try again
-        continue;
+        m_parent.m_maker->operator()(in, out_vec);
+      } catch (...) { // TODO BJL May 28-2021 can we restrict the possible exceptions triggeralgs might raise?
+        ers::fatal(AlgorithmFatalError(ERS_HERE, m_parent.get_name(), m_parent.m_algorithm_name));
+        return;
+      }
+
+      while (out_vec.size() && running_flag.load()) {
+        if (m_parent.send(out_vec.back())) {
+          out_vec.pop_back();
+        }
       }
-      ++received_count;
+    } // end while (running_flag.load())
 
-      std::vector<OUT> outs;
-      m_maker->operator()(in, outs);
+    TLOG() << ": Exiting do_work() method, received " << m_parent.m_received_count 
+           << " inputs and successfully sent " << m_parent.m_sent_count << " outputs. ";
+  }
+};
 
-      while (outs.size()) {
-        bool successfullyWasSent = false;
-        while (!successfullyWasSent && running_flag.load()) {
+// Partial specilization for IN = Set<A>, OUT = Set<B> and assumes the MAKER has:
+// operator()(A, std::vector<B>)
+template<class A, class B, class MAKER> 
+class TriggerGenericWorker<Set<A>, Set<B>, MAKER> 
+{
+public:
+  explicit TriggerGenericWorker(TriggerGenericMaker<Set<A>, Set<B>, MAKER> &parent) : m_parent(parent) { }
+  
+  TriggerGenericMaker<Set<A>, Set<B>, MAKER> &m_parent;
+  
+  void do_work(std::atomic<bool> &running_flag)
+  {
+    while (running_flag.load()) {
+      Set<A> in;
+      if (!m_parent.receive(in)) {
+        continue; //nothing to do
+      }
+
+      Set<B> out; // 1 input set -> 1 output set
+      switch (in.type) {
+        case Set<A>::Type::kPayload:
```

Comment:
I think we need a little more logic here to ensure that `operator()` is called with objects in `time_start` order when the inputs are derived from multiple streams (eg, TAs from multiple APAs)

You can assume that the input queue of sets is time-ordered, one set per stream. That's the way zipper will order the output, ie, like this:

![zipper-output](https://user-images.githubusercontent.com/39594854/120000543-d606c900-bfca-11eb-8b74-e2e9d0f8fcfd.png)

Then you have to "buffer" the input sets until you see one with a larger start time. At that point you can sort the objects from all of the buffered sets and pass them to `operator()`

Reply:
In that case, I'm not sure I understand the purpose of a `Set<T>`, or the logic for how a `TPSet` should map onto a `TASet`. If we're just passing TP / TA to the algorithms in time order, and the mapping from TPSet to TASet is arbitrary, why group them?

Is the idea that a `Set<T>` represents a non-overlapping global time slice? i.e. a maker should wait for all the TPSet from a non-overlapping slice, call the algorithm for all objects in that slice in time order, then is free to aggregate all the results into a single TASet? 

If that's the case, is there a mechanism to ensure the `start_time` and `end_time` of the `Set<T>` in a group is the same that should be preserved through this TP->TA->TC chain?

Reply:
> In that case, I'm not sure I understand the purpose of a `Set<T>`, or the logic for how a `TPSet` should map onto a `TASet`. If we're just passing TP / TA to the algorithms in time order, and the mapping from TPSet to TASet is arbitrary, why group them?

The original motivation for grouping the TPs is as an optimization to reduce message rate across the network: eg, in ProtoDUNE, there are ~1MHz of TPs per APA, so reducing message rate is worthwhile. For TAs, this argument probably doesn't apply, or at least not as strongly, but "packaging" them in the same way as TPs means we can reuse algorithms like the Zipper.

> Is the idea that a `Set<T>` represents a non-overlapping global time slice? i.e. a maker should wait for all the TPSet from a non-overlapping slice, call the algorithm for all objects in that slice in time order, then is free to aggregate all the results into a single TASet?

Yes, that's right. In the diagram I pasted earlier, the values t0, t1, t2 are the same for each of the input streams (which makes things global) and the intervals [t0,t1], [t1,t2] etc are non-overlapping. 

> If that's the case, is there a mechanism to ensure the `start_time` and `end_time` of the `Set<T>` in a group is the same that should be preserved through this TP->TA->TC chain?

The `TPWindow` algorithm being adapted from ptmp (slide 10 of [my talk](https://indico.fnal.gov/event/47067/contributions/215031/attachments/142900/180529/ds-data-transport-2021-05-11.pdf) ) ensures this - it rearranges TPs between TPSets so that the output TPSet widths/offsets are as given in configuration. (Eg, if given width=100, offset=30, the output TPSets would have [start_time, end_time] equal to [30, 129], [130, 229], [230, 339] etc).

Then the big-picture configuration would be that N streams of TPs each go into a TPWindow instance (all the instances have the same width and offset), with the outputs fed into `TPZipper` to merge them in a time-ordered way, and the output of the `TPZipper` fed to `TriggerGenericMaker`

Reply:
Thanks, that makes sense. I'll add some logic to the `TriggerGenericWorker` to buffer, sort, and preserve the time bounds.